File: examples/algorithms-cpp/dynamic.cpp
/**
 * dynamic.cpp - C++动态规划算法实现
 * 包含：0/1背包、LIS长度、零钱问题
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <cassert>

namespace dynamic {

int knapsack01(const std::vector<int>& weights, const std::vector<int>& values, int capacity) {
    int n = static_cast<int>(weights.size());
    std::vector<int> dp(capacity + 1, 0);
    for (int i = 0; i < n; ++i) {
        for (int c = capacity; c >= weights[i]; --c) {
            dp[c] = std::max(dp[c], dp[c - weights[i]] + values[i]);
        }
    }
    return dp[capacity];
}

int lisLength(const std::vector<int>& arr) {
    if (arr.empty()) return 0;
    std::vector<int> tails;
    for (int x : arr) {
        auto it = std::lower_bound(tails.begin(), tails.end(), x);
        if (it == tails.end()) tails.push_back(x);
        else *it = x;
    }
    return static_cast<int>(tails.size());
}

int coinChangeMin(const std::vector<int>& coins, int amount) {
    if (amount == 0) return 0;
    std::vector<int> dp(amount + 1, INT_MAX - 1);
    dp[0] = 0;
    for (int i = 1; i <= amount; ++i) {
        for (int coin : coins) {
            if (coin > 0 && coin <= i && dp[i - coin] != INT_MAX - 1) {
                dp[i] = std::min(dp[i], dp[i - coin] + 1);
            }
        }
    }
    return dp[amount] == INT_MAX - 1 ? -1 : dp[amount];
}

int coinChangeWays(const std::vector<int>& coins, int amount) {
    std::vector<int> dp(amount + 1, 0);
    dp[0] = 1;
    for (int coin : coins) {
        for (int i = coin; i <= amount; ++i) {
            dp[i] += dp[i - coin];
        }
    }
    return dp[amount];
}

} // namespace dynamic

int main() {
    using namespace dynamic;
    assert(knapsack01({2, 3, 4, 5}, {3, 4, 5, 6}, 5) == 7);
    assert(lisLength({10, 9, 2, 5, 3, 7, 101, 18}) == 4);
    assert(lisLength({}) == 0);
    assert(coinChangeMin({1, 2, 5}, 11) == 3);
    assert(coinChangeMin({2}, 3) == -1);
    assert(coinChangeWays({1, 2, 5}, 5) == 4);
    std::cout << "All dynamic programming tests passed!" << std::endl;
    return 0;
}

File: examples/algorithms-js/dynamic.js
/**
 * dynamic.js - JavaScript动态规划算法实现
 * 包含：0/1背包、LIS长度、零钱问题
 */

class Dynamic {
    /**
     * 0/1背包问题
     * @param {number[]} weights - 重量数组
     * @param {number[]} values - 价值数组
     * @param {number} capacity - 容量
     * @returns {number} 最大价值
     */
    static knapsack01(weights, values, capacity) {
        const n = weights.length;
        const dp = new Array(capacity + 1).fill(0);
        for (let i = 0; i < n; i++) {
            for (let c = capacity; c >= weights[i]; c--) {
                dp[c] = Math.max(dp[c], dp[c - weights[i]] + values[i]);
            }
        }
        return dp[capacity];
    }

    /**
     * 最长递增子序列长度
     * @param {number[]} arr
     * @returns {number}
     */
    static lisLength(arr) {
        if (arr.length === 0) return 0;
        const tails = [];
        for (const x of arr) {
            let left = 0, right = tails.length;
            while (left < right) {
                const mid = Math.floor((left + right) / 2);
                if (tails[mid] < x) left = mid + 1;
                else right = mid;
            }
            tails[left] = x;
            if (left === tails.length) tails.push(x);
        }
        return tails.length;
    }

    /**
     * 零钱最小数量
     * @param {number[]} coins
     * @param {number} amount
     * @returns {number} 最小数量，不可能返回 -1
     */
    static coinChangeMin(coins, amount) {
        if (amount === 0) return 0;
        const INF = Number.MAX_SAFE_INTEGER - 1;
        const dp = new Array(amount + 1).fill(INF);
        dp[0] = 0;
        for (let i = 1; i <= amount; i++) {
            for (const coin of coins) {
                if (coin > 0 && coin <= i && dp[i - coin] !== INF) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount] === INF ? -1 : dp[amount];
    }

    /**
     * 零钱组合数
     */
    static coinChangeWays(coins, amount) {
        const dp = new Array(amount + 1).fill(0);
        dp[0] = 1;
        for (const coin of coins) {
            for (let i = coin; i <= amount; i++) {
                dp[i] += dp[i - coin];
            }
        }
        return dp[amount];
    }
}

// 简单测试
function assertEq(name, expected, actual) {
    if (expected !== actual) throw new Error(`FAIL ${name}: expected ${expected}, got ${actual}`);
}

function runTests() {
    assertEq("Knapsack", 7, Dynamic.knapsack01([2, 3, 4, 5], [3, 4, 5, 6], 5));
    assertEq("LIS", 4, Dynamic.lisLength([10, 9, 2, 5, 3, 7, 101, 18]));
    assertEq("LIS empty", 0, Dynamic.lisLength([]));
    assertEq("CoinMin", 3, Dynamic.coinChangeMin([1, 2, 5], 11));
    assertEq("CoinImpossible", -1, Dynamic.coinChangeMin([2], 3));
    assertEq("CoinWays", 4, Dynamic.coinChangeWays([1, 2, 5], 5));
    console.log("All dynamic programming tests passed!");
}

runTests();

module.exports = { Dynamic };
